src/internal/pfsdb/branches.go

```diff
@@ -68,5 +105,173 @@ func UpsertBranch(ctx context.Context, tx *pachsql.Tx, branchInfo *pfs.BranchInf
 	).Scan(&branchID); err != nil {
 		return 0, errors.Wrap(err, "could not create branch")
 	}
+	// Compare current direct provenance to new direct provenance.
+	oldDirectProv, err := GetDirectBranchProvenance(ctx, tx, branchID)
+	if err != nil {
+		return 0, errors.Wrap(err, "could not get direct branch provenance")
+	}
+	// Add net new direct provenance relationships.
+	toAdd := SliceDiff[string, *pfs.Branch](branchInfo.DirectProvenance, oldDirectProv, func(branch *pfs.Branch) string { return branch.Key() })
```

Comment:
This looks correct, but would it be simpler to just delete all entries beforehand and then create new entries. That might be something you could do as 2 total queries. 

It may not matter because I assume the cardinality of branch provenance is fairly low so performance is not a huge concern.

Reply:
Good point, but I think there's a tradeoff here. Although it is simpler to delete all direct provenance relationships and re-create new ones, this does not reflect how pachyderm is generally used. Branch provenance updates usually involve the addition of a pipeline, and creating net new branch provenance relationships.

I think we should favour the common case here, and only execute a database delete when necessary.

However, you did inspire me to batch the insert and deletes, so that we only need two queries. 👍